packages/sanity/playwright-ct/tests/formBuilder/inputs/PortableText/FocusTracking.spec.tsx

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable max-nested-callbacks */
+import {expect, test} from '@playwright/experimental-ct-react'
+import React from 'react'
+import {Path, SanityDocument} from '@sanity/types'
+import {Page} from '@playwright/test'
+import FocusTrackingStory from './FocusTrackingStory'
+
+export type UpdateFn = () => {focusPath: Path; document: SanityDocument}
+
+const document: SanityDocument = {
+  _id: '123',
+  _type: 'test',
+  _createdAt: new Date().toISOString(),
+  _updatedAt: new Date().toISOString(),
+  _rev: '123',
+  body: [
+    {
+      _type: 'block',
+      _key: 'a',
+      children: [{_type: 'span', _key: 'b', text: 'Foo'}],
+      markDefs: [],
+    },
+    {
+      _type: 'block',
+      _key: 'c',
+      children: [{_type: 'span', _key: 'd', text: 'Bar'}],
+      markDefs: [],
+    },
+    {
+      _type: 'block',
+      _key: 'e',
+      children: [{_type: 'span', _key: 'f', text: 'Baz'}],
+      markDefs: [],
+    },
+    {
+      _type: 'block',
+      _key: 'g',
+      children: [
+        {_type: 'span', _key: 'h', text: 'Hello '},
+        {_type: 'inlineObjectWithTextProperty', _key: 'i', text: 'there'},
+        {_type: 'span', _key: 'j', text: ' playwright'},
+      ],
+      markDefs: [],
+    },
+    {
+      _type: 'testObjectBlock',
+      _key: 'k',
+      text: 'Hello world',
+    },
+  ],
+}
+
+test.describe('Portable Text Input', () => {
+  test.beforeEach(async ({page}) => {
+    await page.evaluate(() => {
+      window.localStorage.debug = 'sanity-pte:*'
+    })
+  })
+  test.describe('Should track focusPath', () => {
+    test(`for span .text`, async ({mount, page}) => {
+      const initialPath = ['body', {_key: 'c'}, 'children', {_key: 'd'}, 'text']
+      const component = await mount(
+        <FocusTrackingStory document={document} focusPath={initialPath} />,
+      )
+      await waitForFocusPath(page, initialPath)
+      const $portableTextInput = component.getByTestId('field-body')
+      const $pteTextbox = $portableTextInput.getByRole('textbox')
+      await expect($pteTextbox).toBeFocused()
+      expect(await getFocusedNodeText(page)).toEqual('Bar')
+      await setFocusPathFromOutside(page, ['body', {_key: 'e'}, 'children', {_key: 'f'}, 'text'])
```

Comment:
**question:** instead of using `setFocusPathFromOutside` – would it instead be preferable to re-render the component with new props? e.g.

```
const newPath = ['body', {_key: 'e'}, 'children', {_key: 'f'}, 'text']
await component.update(<FocusTrackingStory focusPath={newPath} />)
expect(await getFocusedNodeText(page)).toEqual('Baz')
```

Reply:
Yes, this is the problem, because they are not synchronous. They are dependent on the editor showing the correct value, before it applies the focusPath (or we will not be able to create the correct window.selection).